pallets/elections/src/lib.rs

```diff
@@ -443,26 +468,46 @@ pub mod pallet {
         type Error = ElectionError;
         type DataProvider = T::DataProvider;
 
-        /// The elections are PoA so only the nodes listed in the Validators will be elected as
-        /// validators.
-        ///
-        /// We calculate the supports for them for the sake of eras payouts.
+        /// We calculate the supports for each validator. The external validators are chosen as:
+        /// 1) "`NextEraNonReservedValidators` that are staking and are not banned" in case of Permissioned ElectionOpenness
+        /// 2) "All staking and not banned validators" in case of Permissionless ElectionOpenness
         fn elect() -> Result<Supports<T::AccountId>, Self::Error> {
-            Self::kick_out_underperformed_non_reserved_validators();
+            Self::emit_fresh_bans_event();
+            let active_era = <T as Config>::EraInfoProvider::active_era().unwrap_or(0);
+            let ban_period = BanConfig::<T>::get().ban_period;
 
             let staking_validators = Self::DataProvider::electable_targets(None)
                 .map_err(Self::Error::DataProvider)?
                 .into_iter()
                 .collect::<BTreeSet<_>>();
-            let reserved_validators = NextEraReservedValidators::<T>::get()
+            let staking_reserved_validators = NextEraReservedValidators::<T>::get()
                 .into_iter()
+                .filter(|v| staking_validators.contains(v))
+                .collect::<BTreeSet<_>>();
+            let banned_validators = Banned::<T>::iter()
+                .filter(|(_, info)| !Self::ban_expired(info.start, ban_period, active_era + 1))
+                .map(|(v, _)| v)
                 .collect::<BTreeSet<_>>();
-            let non_reserved_validators = NextEraNonReservedValidators::<T>::get()
+            let old_non_reserved_validators = NextEraNonReservedValidators::<T>::get().into_iter();
+
+            let eligible_non_reserved = staking_validators
                 .into_iter()
+                .filter(|v| {
+                    !banned_validators.contains(v) && !staking_reserved_validators.contains(v)
+                })
                 .collect::<BTreeSet<_>>();
 
-            let eligible_validators =
-                &(&reserved_validators | &non_reserved_validators) & &staking_validators;
+            let new_non_reserved_validators: Vec<_> = match Openness::<T>::get() {
+                ElectionOpenness::Permissioned => old_non_reserved_validators
+                    .filter(|v| eligible_non_reserved.contains(v))
+                    .collect(),
+                ElectionOpenness::Permissionless => eligible_non_reserved.into_iter().collect(),
+            };
+            NextEraNonReservedValidators::<T>::put(new_non_reserved_validators.clone());
```

Comment:
I think it would be great to mention here in a comment why we change it. This is because we need the following invariant to hold here:
`eligible_validators ` (i.e. what we output from `elect()`) should be the disjoint union of `NextEraNonReservedValidators` + `NextEraReservedValidators`.

This is because of what happens in `populate_next_era_validators_on_next_era_start` which is performed in the same block, right after `elect`.

Reply:
added.